content/lectures/lec02.md

```diff
@@ -7,3 +7,190 @@ draft: false
+test command: 
+git repository: 
+keywords: 
+author: 
+license: (ISC) 
+About to write to /Users/campbellphalen/Desktop/node_demo/package.json:
+
+{
+  "name": "node_demo",
+  "version": "1.0.0",
+  "description": "A demo application built to outline development with Docker for cis188",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1"
+  },
+  "author": "",
+  "license": "ISC"
+}
+
+
+Is this OK? (yes) yes
+```
+
+This will create a `package.json` file containing the information we've just provided. Finally, we need our `index.js` file:
+
+```
+const http = require('http');
+
+const hostname = '0.0.0.0';
+const port = 8080;
+
+const server = http.createServer((req, res) => {
+  res.statusCode = 200;
+  res.setHeader('Content-Type', 'text/plain');
+  res.end('Hello World!');
+});
+
+server.listen(port, hostname, () => {
+  console.log(`Server running at http://${hostname}:${port}/`);
+});
+```
+
+Finally, let's build the Docker image and start up the Docker container:
+
+```
+# Build the Docker image from our Dockerfile
+$ docker build --tag node-demo-image .
+# Run the Docker container with port 8080 exposed
+$ docker run -p 8080:8080 --name node-demo-container node-demo-image
+Server running at http://0.0.0.0:8080/
+```
+
+Now, when we navigate to (http://localhost:8080/)[http://localhost:8080/] we can see the `Hello world!` message. So, let's make some changes to this web application and see what we need to do in our Dockerfile to support those changes (spoiler alert: no very much, Docker makes the development process pretty seemless).
+
+Let's say we needed to generate unique user IDs if we were saving users into a database. There's an Node package called `uuid`, let's install that and see how Docker handles adding a new package. We can type:
+
+```
+$ npm install uuid
+```
+
+then, let's modify line 10 in our `index.js` file form
+
+```
+res.end('Hello World!');
+```
+
+to be
+
+```
+res.end('Hello World: ' + uuidv4());
+```
+
+and also put this import statement at the top of the file
+
+```
+const { v4: uuidv4 } = require('uuid');
+```
+
+then, when we re-build our Docker container we will see this new Node package
+installed, and our index page will be modified:
+
+```
+# Delete the previous node-demo container (cannot have duplicate container names)
+$ docker container rm node-demo-container
+# Rebuild the node-demo-image
+$ docker build --tag node-demo-image .
+# Recreate the node-demo container
+$ docker run -p 8080:8080 --name node-demo-container node-demo-image
+```
+
+Now, this is pretty concise. When we make a change to our development environment (here we installed a new dependency), Docker makes it pretty seemless to also have that be installed the Docker container's environment. However, we still have to rebuild the Docker image and the Docker container, which can feel a little tedious. This is something we'll address later in the course when we talk about Docker-compose!
```

Comment:
seamless. also compose handles runtime config, but it doesn't handle this rebuilding at all. should probably remove that